packages/next/build/webpack/plugins/next-trace-entrypoints-plugin.ts

```diff
@@ -0,0 +1,239 @@
+      compilation.hooks.finishModules.tapAsync(
+        PLUGIN_NAME,
+        async (_stats: any, callback: any) => {
+          // we create entry -> module maps so that we can
+          // look them up faster instead of having to iterate
+          // over the compilation modules list
+          const entryNameMap = new Map<string, string>()
+          const entryModMap = new Map<string, any>()
+
+          try {
+            const depModMap = new Map<string, any>()
+
+            compilation.entries.forEach((entry) => {
+              const name = entry.name || entry.options?.name
+
+              if (name?.startsWith('pages/') && entry.dependencies[0]) {
+                const entryMod = getModuleFromDependency(
+                  compilation,
+                  entry.dependencies[0]
+                )
+
+                if (entryMod.resource) {
+                  entryNameMap.set(entryMod.resource, name)
+                  entryModMap.set(entryMod.resource, entryMod)
+                }
+              }
+            })
+
+            // TODO: investigate allowing non-sync fs calls in node-file-trace
+            // for better performance
+            const readFile = (path: string) => {
+              const mod = depModMap.get(path) || entryModMap.get(path)
+
+              // map the transpiled source when available to avoid
+              // parse errors in node-file-trace
+              const source = mod?.originalSource()
+
+              if (source?._valueAsBuffer) {
+                return source._valueAsBuffer
+              }
+
+              try {
+                return compilation.inputFileSystem.readFileSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT' || e.code === 'EISDIR') {
+                  return null
+                }
+                throw e
+              }
+            }
+            const readlink = (path: string) => {
+              try {
+                return compilation.inputFileSystem.readlinkSync(path)
+              } catch (e) {
+                if (
+                  e.code !== 'EINVAL' &&
+                  e.code !== 'ENOENT' &&
+                  e.code !== 'UNKNOWN'
+                ) {
+                  throw e
+                }
+                return null
+              }
+            }
+            const stat = (path: string) => {
+              try {
+                return compilation.inputFileSystem.statSync(path)
+              } catch (e) {
+                if (e.code === 'ENOENT') {
+                  return null
+                }
+                throw e
+              }
+            }
+
+            const nftCache = {}
+            const entryPaths = Array.from(entryModMap.keys())
+
+            for (const entry of entryPaths) {
+              depModMap.clear()
+              const entryMod = entryModMap.get(entry)
+              const cachedTraces = entryMod.buildInfo?.cachedNextEntryTrace
```

Comment:
That's too naive caching. Even if the module is unchanged it could be that files that are only read by nft have changed. It's not safe that way.

I would omit that for this PR and add caching in a separate PR.

We would need to track all files accessed by nft (good that it exposes the fs access methods), create a snapshot from that, store the traces and the snapshot and validate the snapshot on restore.

Reply:
Commented out the caching for now noting why it's disabled, will investigate it more in a follow-up PR